client/src/main/java/com/tigrisdata/db/client/service/TigrisAsyncCollection.java

```diff
@@ -0,0 +1,161 @@
+      TigrisDBAsyncReader<T> reader)
+      throws TigrisDBException;
+
+  /**
+   * Reads matching documents
+   *
+   * @param filter filter to narrow down read
+   * @param fields optionally specify fields you want to be returned from server
+   * @param reader reader callback
+   * @throws TigrisDBException in case of an error
+   */
+  void read(TigrisFilter filter, List<Field<?>> fields, TigrisDBAsyncReader<T> reader)
+      throws TigrisDBException;
+
+  /**
+   * Reads a single document. This method is generally recommended for point lookup, if used for
+   * non-point lookup any arbitrary matching document will be returned.
+   *
+   * @param filter
+   * @return a future to the document
+   * @throws TigrisDBException
+   */
+  CompletableFuture<T> readOne(TigrisFilter filter) throws TigrisDBException;
+
+  /**
+   * @param documents list of documents to insert
+   * @param insertRequestOptions insert option
+   * @return a future to the {@link InsertResponse}
+   * @throws TigrisDBException in case of an error
+   */
+  CompletableFuture<InsertResponse> insert(
+      List<T> documents, InsertRequestOptions insertRequestOptions) throws TigrisDBException;
+
+  /**
+   * @param documents list of documents to insert
+   * @return a future to the {@link InsertResponse}
+   * @throws TigrisDBException in case of an error
+   */
+  CompletableFuture<InsertResponse> insert(List<T> documents) throws TigrisDBException;
+
+  /**
+   * inserts a single document to the collection
+   *
+   * @param document
+   * @return a future to the {@link InsertResponse}
+   * @throws TigrisDBException
+   */
+  CompletableFuture<InsertResponse> insert(T document) throws TigrisDBException;
+
+  /**
+   * Inserts the documents if they don't exist already (matched by primary key), replaces them
+   * otherwise.
+   *
+   * @param documents list of documents to replace
+   * @param insertOrReplaceRequestOptions replace option
+   * @return a future to the {@link InsertOrReplaceResponse}
+   * @throws TigrisDBException in case of an error
+   */
+  CompletableFuture<InsertOrReplaceResponse> insertOrReplace(
+      List<T> documents, InsertOrReplaceRequestOptions insertOrReplaceRequestOptions)
+      throws TigrisDBException;
+
+  /**
+   * Inserts the documents if they don't exist already (matched by primary key), replaces them
+   * otherwise.
+   *
+   * @param documents list of documents to replace
+   * @return a future to the {@link InsertOrReplaceResponse}
+   * @throws TigrisDBException in case of an error
+   */
+  CompletableFuture<InsertOrReplaceResponse> insertOrReplace(List<T> documents)
+      throws TigrisDBException;
+
+  /**
+   * @param filter
+   * @param fields
+   * @param updateRequestOptions
+   * @return a future to the {@link UpdateResponse}
+   * @throws TigrisDBException
+   */
+  CompletableFuture<UpdateResponse> update(
+      TigrisFilter filter, List<Field<?>> fields, UpdateRequestOptions updateRequestOptions)
```

Comment:
is this Field class same as read? They should be different.

Reply:
I will refactor this in next PR.